### DISCUSSION TITLE
[question] run commands after start and before stop

### DISCUSSION THREAD

**Discussion by dkwo** [davmac314/dinit#419] [General]
Perhaps it is already possible, but I am struggling to implement a user service that runs a daemon like so:
```
type = process
env-file = env-var-xdg
command = /usr/bin/pizauth server -vvvv -d
stop-command = /usr/bin/pizauth shutdown
log-type = buffer
```
where I would like to run commands
```
age -d -i $PRIVKEY -o - ${XDG_STATE_HOME}/pizauth.age | pizauth restore
pizauth dump | age -e -R $PUBKEY -o ${XDG_STATE_HOME}/pizauth.age
```
after it starts and before it stops, respectively.

My attempt does not seem to work:
```
type = process
env-file = env-var-xdg
load-options = export-passwd-vars
command = /usr/bin/pizauth server -vvvv -d
command += sh -c "age -d -i ${HOME}/.passage/identities -o - ${XDG_STATE_HOME}/pizauth.age | pizauth restore"
stop-command = sh -c "pizauth dump | age -e -r $pubkey -o ${XDG_STATE_HOME}/pizauth.age"
stop-command += /usr/bin/pizauth shutdown
log-type = buffer
```
Do you know a way to fix this? Or would it make sense to implement this as a new feature, akin to systemd `ExecStartPost`?
Thanks!

**Comment by davmac314** [comment]
The `command +=` and `stop-command +=` are adding to the existing command/stop command, not adding additional commands. Each service can only have one associated process (command).

What you need is two services, one which depends on the other.

Eg (untested):

Service "pizauth-daemon", almost as you wrote above:
```
type = bgprocess
env-file = env-var-xdg
command = /usr/bin/pizauth server -vvvv
stop-command = /usr/bin/pizauth shutdown
log-type = buffer
```

I used `type = bgprocess` and removed `-d` from the command line, because I don't think pizauth supports any readiness notification. Without that then you have a race between starting the daemon and trying to connect to it via `pizauth restore`. An alternative could be to insert a delay before `pizauth restore` (`sh -c "sleep 1; age -d -i ... (etc)`).

Service "pizauth"
```
type = scripted
env-file = env-var-xdg
load-options = export-passwd-vars
command = sh -c "age -d -i ${HOME}/.passage/identities -o - ${XDG_STATE_HOME}/pizauth.age | pizauth restore"
stop-command = sh -c "pizauth dump | age -e -r $pubkey -o ${XDG_STATE_HOME}/pizauth.age"
log-type = buffer
depends-on: pizauth-daemon
```

Now when you start "pizauth", it first starts "pizauth-daemon" (which runs `/usr/bin/pizauth server -vvvv`) and then runs `sh -c "age -d -i ${HOME}/.passage/identities -o - ${XDG_STATE_HOME}/pizauth.age | pizauth restore"`.

When you stop "pizauth", it runs `sh -c "pizauth dump | age -e -r $pubkey -o ${XDG_STATE_HOME}/pizauth.age"`, then stops `pizauth-daemon`, which runs `/usr/bin/pizauth shutdown`.

**Reply by dkwo** [reply]
Thanks a lot! I was not sure what to specify for the bgprocess pid-file, so I went with the `sleep 1` option for now. I will test it for a while. Also, I had to use `/bin/sh -c`.

**Reply by davmac314** [reply]
Ah true, pizauth doesn't seem to write a PID file. You could also use a `type = scripted` service for the daemon itself (but you would not get supervision in that case; if the process terminated, dinit wouldn't know).
